### DISCUSSION TITLE
Wrong yarn version being used

### DISCUSSION THREAD

**Discussion by adamwilson-mse** [renovatebot/renovate#11646] [Request Help] [answered]
Hi, having some problems trying to use an older version of yarn while self-hosting renovate bot. We need to use yarn v1.19 because of our project using some older packages which have some deprecated syntax (yes we will fix this soon but can't yet!).

We are using GitLab and the renovate project we are using was forked from https://gitlab.com/gitlab-org/frontend/renovate-gitlab-bot.

The renovate version we are using is 26.21.7.

No matter what we try, the "yarn install --ignore-engines --ignore-platform --network-timeout 100000" command that renovate runs on the branch of the target repository is using yarn v1.22.5.

Dockerfile to build renovate image:
```
FROM node:12.22.6-alpine3.12 AS builder

ADD / /workdir
WORKDIR /workdir

RUN apk add --update --no-cache python3 make g++

RUN yarn install --production --frozen-lockfile --force

FROM ruby:2.6.6-alpine3.12

COPY --from=builder /workdir /workdir
COPY --from=builder /usr/local/bin/node /usr/local/bin/node
COPY --from=builder /usr/local/lib/node_modules /usr/local/lib/node_modules
COPY --from=builder /usr/local/include/node /usr/local/lib/node
COPY --from=builder /opt/yarn-* /opt/yarn

ENV PATH=/opt/yarn/bin:${PATH}

RUN  apk add --update --no-cache git bash \
  && node -v && cd /usr/local/bin/ \
  && ln -s ../lib/node_modules/npm/bin/npm-cli.js /usr/local/bin/npm \
  && ln -s ../lib/node_modules/npm/bin/npx-cli.js /usr/local/bin/npx \
  && ln -s /opt/yarn/bin.yarn /usr/local/bin/yarn \
  && yarn -v && npm -v && npx -v

WORKDIR /workdir
```

We have tried:
- Installing yarn v1.19 in the Dockerfile that builds the renovate image (we later learned that this yarn version is not used for yarn install, as shown [here](https://github.com/renovatebot/renovate/blob/main/lib/manager/npm/post-update/yarn.ts#L68)?)
- Adding "constraints" to renovate section in package.json on the target repo
- Adding "constraints" to config.js on the renovate repo

The constraints section we've been using looks like this:
```
    "force": {
      "constraints": {
        "yarn": "1.19.1"
      }
    }
```

From the debug logs renovate does read this constraint but doesn't seem to then use 1.19.1:
```
DEBUG: Using package.json for global renovate config (repository=<redacted>)
DEBUG: Found package.json config file (repository=<redacted>)
DEBUG: package.json>renovate config (repository=<redacted>)
       "config": {
         "force": {"constraints": {"yarn": "1.19.1"}}
       }
DEBUG: migrateAndValidate() (repository=<redacted>)
DEBUG: No config migration necessary (repository=<redacted>)
DEBUG: massaged config (repository=<redacted>)
       "config": {
         "force": {"constraints": {"yarn": "1.19.1"}}
       }
DEBUG: migrated config (repository=<redacted>)
       "config": {
         "force": {"constraints": {"yarn": "1.19.1"}}
       }
```

```
DEBUG: rawExec err (repository=<redacted>, branch=renovate/jest-monorepo)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "yarn install --ignore-engines --ignore-platform --network-timeout 100000",
         "stdout": "yarn install v1.22.5\n[1/4] Resolving packages...\n[2/4] Fetching packages...\n[3/4] Linking dependencies...\n[4/4] Building fresh packages...
```

Please let me know if any other info would be useful.

Has anyone had any success in specifying which yarn version to use? Is constraints the right config option to use?

Any help would be much appreciated! Thank you!

**Comment by viceice** [comment, marked as answer]
You need to run renovate in docker mode

https://gitlab.com/renovate-bot/renovate-runner

Or you need to have your specific yarn version globally installed. When you are using default global binarySource.

**Reply by adamwilson-mse** [reply]
Thanks for the quick reply! How would we install it globally for renovate to use? Changing our Dockerfile to this successfully changes the yarn version for the yarn install command after, but when renovate runs using this image in the next CI job it still uses v1.22.5 :(

```
FROM node:12.22.6-alpine3.12 AS builder

ADD / /workdir
WORKDIR /workdir

RUN apk add --update --no-cache python3 make g++

# Downgrade yarn for installing old npm modules
RUN npm install --global --force yarn@1.19.1

RUN yarn install --production --frozen-lockfile --force

FROM ruby:2.6.6-alpine3.12

COPY --from=builder /workdir /workdir
COPY --from=builder /usr/local/bin/node /usr/local/bin/node
COPY --from=builder /usr/local/lib/node_modules /usr/local/lib/node_modules
COPY --from=builder /usr/local/include/node /usr/local/lib/node
COPY --from=builder /opt/yarn-* /opt/yarn

ENV PATH=/opt/yarn/bin:${PATH}

RUN  apk add --update --no-cache git bash \
  && node -v && cd /usr/local/bin/ \
  && ln -s ../lib/node_modules/npm/bin/npm-cli.js /usr/local/bin/npm \
  && ln -s ../lib/node_modules/npm/bin/npx-cli.js /usr/local/bin/npx \
  && ln -s /opt/yarn/bin.yarn /usr/local/bin/yarn \
  && yarn -v && npm -v && npx -v

WORKDIR /workdir
```

**Reply by adamwilson-mse** [reply]
Thanks @viceice managed to get this working by using a combination of the "docker" binarySource, adding docker to the renovate Dockerfile, and having the constraints section in our renovate config.
